File: Cake/reverseString.js

function reverseString(str) {
    var str_list = str.split(""),
        lo = 0,
        hi = str_list.length - 1;

    while (lo < hi) {
        var temp = str_list[lo];
        str_list[lo] = str_list[hi];
        str_list[hi] = temp;
        lo++;
        hi--;
    }
    return str_list.join("");
}

var str = "hello word";
console.log(str);
var answer = reverseString(str);
console.log(answer);

File: Cake/reverseString.py
"""
    Reverse a string in place.
    Strings are immutable in Python. Convert to a list first. Swap values
    at low index with values at high index, coming towards the middle.
    O(n) time
    O(1) space
"""

class Solution(object):
    def reverse_string(self, str):
        str_list = list(str)
        lo = 0
        hi = len(str_list) - 1

        while lo < hi:
            str_list[lo], str_list[hi] = str_list[hi], str_list[lo]
            lo += 1
            hi -= 1

        return "".join(str_list)

if __name__ == "__main__":
    s = Solution()
    str = "hello world to you"
    print(str)
    answer = s.reverse_string(str)
    print(answer)
